Mathlib/NumberTheory/LSeries/ZMod.lean

```diff
@@ -152,49 +183,382 @@ lemma LFunction_stdAddChar_eq_expZeta (j : ZMod N) (s : ℂ) (hjs : j ≠ 0 ∨
+  simp only [LFunction_def_odd hΦ, hurwitzZetaOdd_neg_two_mul_nat_sub_one, mul_zero, sum_const_zero]
+
+variable (Φ) in
+/--
+The completed L-function of a function mod `N`.
+
+This is only mathematically meaningful if `Φ` is either even, or odd. We extend this to all `Φ`
+as follows: if `Φ` is not even, then the completed L-series of `Φ` is the `L`-series of the odd
+part of `Φ`.
+-/
+noncomputable def completedLFunction (s : ℂ) : ℂ :=
+  if Φ.Even then N ^ (-s) * ∑ j, Φ j * completedHurwitzZetaEven (toAddCircle j) s
+  else N ^ (-s) * ∑ j, Φ j * completedHurwitzZetaOdd (toAddCircle j) s
+
+lemma completedLFunction_zero (s : ℂ) : completedLFunction (0 : ZMod N → ℂ) s = 0 := by
+  simp only [completedLFunction, Pi.zero_apply, zero_mul, sum_const_zero, mul_zero, ite_self]
+
+lemma completedLFunction_def_even (hΦ : Φ.Even) (s : ℂ) :
+    completedLFunction Φ s = N ^ (-s) * ∑ j, Φ j * completedHurwitzZetaEven (toAddCircle j) s := by
+  simp only [completedLFunction, hΦ, ↓reduceIte]
+
+lemma completedLFunction_def_odd (hΦ : Φ.Odd) (s : ℂ) :
+    completedLFunction Φ s = N ^ (-s) * ∑ j, Φ j * completedHurwitzZetaOdd (toAddCircle j) s := by
+  by_cases hΦ' : Φ.Even
+    -- if `Φ = 0` we are in the "wrong" branch of the if-then, but it's OK because both sides are 0
+  · simp only [Φ.zero_of_even_and_odd hΦ' hΦ, completedLFunction_zero, Pi.zero_apply, zero_mul,
+      sum_const_zero, mul_zero]
+  · simp only [completedLFunction, hΦ', ↓reduceIte]
+
+/--
+The completed L-function of a function mod 1 is a scalar multiple of the completed Riemann zeta
+function.
+-/
+lemma completedLFunction_modOne_eq {Φ : ZMod 1 → ℂ} (s : ℂ) :
+    completedLFunction Φ s = Φ 1 * completedRiemannZeta s := by
+  have : Φ.Even := fun j ↦ congr_arg Φ <| (Unique.eq_default (-j)).trans (Unique.eq_default j).symm
+  simp [completedLFunction, this]
+  change Φ 1 * completedHurwitzZetaEven (toAddCircle 0) s = _
+  rw [map_zero, completedHurwitzZetaEven_zero]
+
+variable (Φ) in
+/--
+The completed L-function of a function mod `N`, modified by adding multiples of `N ^ (-s) / s` and
+`N ^ (-s) / (1 - s)` to make it entire.
+
+This is well-defined for all `Φ`, but is uninteresting unless `Φ` is either even or odd.
+-/
+noncomputable def completedLFunction₀ (s : ℂ) : ℂ :=
+  if Φ.Even then N ^ (-s) * ∑ j : ZMod N, Φ j * completedHurwitzZetaEven₀ (toAddCircle j) s
+  else N ^ (-s) * ∑ j : ZMod N, Φ j * completedHurwitzZetaOdd (toAddCircle j) s
+
+variable (Φ) in
+/-- The function `completedLFunction₀ Φ` is differentiable. -/
+lemma differentiable_completedLFunction₀ : Differentiable ℂ (completedLFunction₀ Φ) := by
+  unfold completedLFunction₀
+  by_cases h : Φ.Even <;>
+  simp only [h, reduceIte] <;>
+  refine (differentiable_neg.const_cpow <| .inl <| NeZero.ne _).mul (.sum fun i _ ↦ .const_mul ?_ _)
+  exacts [differentiable_completedHurwitzZetaEven₀ _, differentiable_completedHurwitzZetaOdd _]
+
+lemma completedLFunction_eq (hΦ : Φ.Even ∨ Φ.Odd) (s : ℂ) :
```

Comment:
We usually state two lemmas instead of one using an `Or`, but I don't think this is too strict, especially if it gives a lot of code duplication (same below).

Reply:
I thought about this, and realised that the duplication would not actually be so bad, because the proof is far easier and shorter for odd functions than even ones (you don't have to mess around so much with possible poles). So I have split it into two separate lemmas for the odd / even cases.